prboom2/src/dsda/hud_components/weapon_text.c

```diff
@@ -20,32 +20,56 @@
 #include "weapon_text.h"
 
 static dsda_text_t component;
+static dboolean grid;
 
 static void dsda_UpdateComponentText(char* str, size_t max_size) {
   player_t* player;
 
   player = &players[displayplayer];
 
-  snprintf(
-    str,
-    max_size,
-    "WPN \x1b%c%c\x1b%c %c %c %c %c %c %c %c %c",
-    player->powers[pw_strength] ? HUlib_Color(CR_BLUE) : HUlib_Color(CR_GREEN),
-    player->weaponowned[0] ? '1' : ' ',
-    HUlib_Color(CR_GREEN),
-    player->weaponowned[1] ? '2' : ' ',
-    player->weaponowned[2] ? '3' : ' ',
-    player->weaponowned[3] ? '4' : ' ',
-    player->weaponowned[4] ? '5' : ' ',
-    player->weaponowned[5] ? '6' : ' ',
-    player->weaponowned[6] ? '7' : ' ',
-    player->weaponowned[7] ? '8' : ' ',
-    player->weaponowned[8] ? '9' : ' '
-  );
+  if (grid)
+    snprintf(
+      str,
+      max_size,
+      "W \x1b%c%c\x1b%c %c %c\n\x1b%cP\x1b%c %c %c %c\n\x1b%cN\x1b%c %c %c %c",
```

Comment:
Let's make this line slightly more digestible by breaking it up. I believe this is the syntax:
```c
"line 1\n"
"line 2\n"
"line 3",
```
If you're feeling fancy you could see how it looks to include the color reversion at the tail of the line instead of the start of the next line, so the W P N are all lined up at the start of the substrings.